File: apps/sim/scripts/probes/session-process.py
import concurrent.futures, os, pathlib, pwd, signal, subprocess, time

assert os.getpid() == 1 and os.getuid() == 0, 'Run only as PID 1 in a disposable root container'

subprocess.run(['useradd', '--create-home', 'user'], check=True)
owner = pwd.getpwnam('user')
base = pathlib.Path('/tmp/mship-kernel-proof')
base.mkdir(mode=0o777)
base.chmod(0o777)
processes = []
def run(job):
    process = subprocess.Popen(['/bin/bash', '-c', job['run']], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    processes.append(process)
    return process
def stop(job):
    result = subprocess.run(['/bin/bash', '-c', job['stop']], capture_output=True, timeout=5)
    assert result.returncode == 0, result.stderr.decode()
    assert json.loads(result.stdout) == {'settled': True}
def size(name):
    path = base / name
    return path.stat().st_size if path.exists() else 0
def started(name):
    for _ in range(300):
        if size(name):
            return
        time.sleep(0.01)
    raise AssertionError('writer did not start: ' + name)

try:
    if expect_no_namespaces:
        refused = run(jobs['first'])
        output, error = refused.communicate(timeout=3)
        assert refused.returncode != 0 and b'Operation not permitted' in error, error
        stop(jobs['first'])
        assert size('first') == 0
        print(json.dumps({'ok': True, 'missingNamespaceCapabilityRefusesUserCode': True}))
        raise SystemExit(0)
    stop(jobs['late'])
    late = run(jobs['late'])
    late.communicate(timeout=3)
    assert late.returncode != 0 and size('late') == 0
    first, sibling = run(jobs['first']), run(jobs['sibling'])
    started('first')
    started('sibling')
    assert (base / 'first').stat().st_uid == owner.pw_uid
    duplicate = run(jobs['first'])
    duplicate.communicate(timeout=3)
    assert duplicate.returncode != 0
    stop(jobs['first'])
    first.communicate(timeout=3)
    stopped_count, sibling_count = size('first'), size('sibling')
    time.sleep(0.08)
    assert size('first') == stopped_count and size('sibling') > sibling_count
    assert (base / 'first').read_text() == 'x' * stopped_count
    for name in ['empty', 'unicode', 'root', 'normal-exit', 'reap-orphans']:
        normal = run(jobs[name])
        output, error = normal.communicate(timeout=3)
        assert normal.returncode == 0, error.decode()
        stop(jobs[name])
        if name == 'unicode':
            assert output.decode() == 'héllo 世界\n/home/user|user|/home/user', output
        if name == 'root':
            assert output.strip() == b'0'
        if name == 'normal-exit':
            count = size(name)
            time.sleep(0.08)
            assert size(name) == count

    launcher = run(jobs['launcher-death'])
    started('launcher-death')
    launcher.kill()
    launcher.communicate(timeout=3)
    count = size('launcher-death')
    time.sleep(0.08)
    assert size('launcher-death') == count
    stop(jobs['launcher-death'])

    controller = os.fork()
    if controller == 0:
        subprocess.Popen(['/bin/bash', '-c', jobs['controller-death']['run']],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        os._exit(0)
    os.waitpid(controller, 0)
    started('controller-death')
    stop(jobs['controller-death'])
    count = size('controller-death')
    time.sleep(0.08)
    assert size('controller-death') == count

    # A recycled host PID must not kill an unrelated running command.
    sibling_state = json.loads(pathlib.Path('/run/sim-mship-processes/v1', jobs['sibling']['id'] + '.json').read_text())
    stale = pathlib.Path('/run/sim-mship-processes/v1', jobs['stale-pid']['id'] + '.json')
    stale.write_text(json.dumps(dict(sibling_state, start='0')))
    count = size('sibling')
    stop(jobs['stale-pid'])
    time.sleep(0.08)
    assert size('sibling') > count

    for index in range(20):
        name = 'race' + str(index)
        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as pool:
            pending_run = pool.submit(run, jobs[name])
            stopped = pool.submit(stop, jobs[name])
            stopped.result()
            observed = size(name)
            time.sleep(0.02)
            assert size(name) == observed, 'execution started after Stop acknowledgement'
            pending_run.result().communicate(timeout=3)
    print(json.dumps({'ok': True, 'actualModuleCommands': True, 'lateLaunchRefused': True,
        'concurrentStopLaunchRaces': 20, 'duplicateLaunchRefused': True,
        'detachedDescendantStopped': True, 'parallelSiblingContinues': True,
        'normalUserFileOwnership': True, 'filesPreserved': True, 'launcherDeathKillsDescendants': True,
        'controllerDeathRecovery': True, 'stalePidDoesNotKillSibling': True,
        'emptyAndUnicodeCommands': True, 'homeUserCwdPreserved': True, 'trustedRootCommand': True,
        'normalExitKillsDetachedChildren': True, 'orphanedChildrenReaped': True}))
finally:
    for job in jobs.values():
        stop(job)
    for process in processes:
        process.communicate(timeout=3)

File: apps/sim/scripts/probes/session-process.ts
/**
 * Emits the production-command Linux probe for an isolated container. Pipe to Python as PID 1
 * in a disposable, network-disabled container with SYS_ADMIN and namespace-capable seccomp.
 * No provider credentials, services, or host filesystem mounts are needed.
 */
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import {
  sessionProcessCommand,
  sessionProcessStopCommand,
} from '@/lib/execution/remote-sandbox/session-process'

const writer = `import os, pathlib, sys, time
path = pathlib.Path(sys.argv[1])
child = os.fork()
if child == 0:
    os.setsid()
    for i in range(1000):
        with path.open('a') as target:
            target.write('x')
        time.sleep(0.01)
    os._exit(0)
print(os.getuid(), flush=True)
time.sleep(10)
`
const quote = (value: string) => `'${value.replaceAll("'", "'\\''")}'`
const jobs = Object.fromEntries(
  [
    'first',
    'sibling',
    'late',
    'launcher-death',
    'controller-death',
    'normal-exit',
    'reap-orphans',
    'stale-pid',
    'empty',
    'unicode',
    'root',
    ...Array.from({ length: 20 }, (_, index) => `race${index}`),
  ].map((name, index) => {
    const id = `00000000-0000-4000-8000-${String(index).padStart(12, '0')}`
    let command = `python3 -c ${quote(writer)} /tmp/mship-kernel-proof/${name}`
    if (name === 'normal-exit')
      command = `python3 -c ${quote(writer.replace('time.sleep(10)', 'time.sleep(0.1)'))} /tmp/mship-kernel-proof/${name}`
    if (name === 'reap-orphans')
      command = `python3 -c ${quote(`import os, pathlib, time
for _ in range(20):
    child = os.fork()
    if child == 0:
        if os.fork() == 0: os._exit(0)
        os._exit(0)
    os.waitpid(child, 0)
for _ in range(100):
    zombies = 0
    for path in pathlib.Path('/proc').glob('[0-9]*/stat'):
        try: zombies += path.read_text().rsplit(')', 1)[1].split()[0] == 'Z'
        except FileNotFoundError: pass
    if not zombies: break
    time.sleep(0.01)
assert zombies == 0, 'namespace init did not reap orphaned children'
`)}`
    if (name === 'empty') command = ''
    if (name === 'unicode')
      command = 'printf \'%s\' \'héllo 世界\'; printf \'\\n%s|%s|%s\' "$HOME" "$USER" "$PWD"'
    if (name === 'root') command = 'id -u'
    return [
      name,
      {
        id,
        run: sessionProcessCommand(id, command, name === 'root'),
        stop: sessionProcessStopCommand(id),
      },
    ]
  })
)
const probeMode = process.argv.includes('--no-namespaces')
const proof = `import json\nexpect_no_namespaces = ${probeMode ? 'True' : 'False'}\njobs = json.loads(${JSON.stringify(JSON.stringify(jobs))})\n${readFileSync(fileURLToPath(new URL('./session-process.py', import.meta.url)), 'utf8')}`
process.stdout.write(proof)
